sktime/benchmarking/benchmarking_dataclasses.py

```diff
@@ -0,0 +1,430 @@
+    dataset_loader: Callable
+        A function which returns a dataset, like from `sktime.datasets`.
+    cv_splitter: BaseSplitter object
+        Splitter used for generating validation folds.
+    scorers: list of BaseMetric objects
+        Each BaseMetric output will be included in the results.
+    """
+
+    id: str
+    dataset_loader: Callable
+    cv_splitter: BaseSplitter
+    scorers: list[BaseMetric]
+
+
+@dataclass
+class ModelToTest:
+    """
+    A model to test.
+
+    Parameters
+    ----------
+    id: str
+        The ID of the model.
+    model: BaseEstimator
+        The model to test.
+    """
+
+    id: str
+    model: BaseEstimator
+
+
+@dataclass
+class ScoreResult:
+    """
+    The result of a single scorer.
+
+    Parameters
+    ----------
+    name: str
+        The name of the scorer.
+    score: float
+        The score.
+    """
+
+    name: str
+    score: float
+
+
+@dataclass
+class FoldResults:
+    """
+    Results for a single fold.
+
+    Parameters
+    ----------
+    fold: int
+        The fold number.
+    scores: list of ScoreResult
+        The scores for this fold for each scorer.
+    ground_truth: pd.Series, optional (default=None)
+        The ground truth series for this fold.
+    predictions: pd.Series, optional (default=None)
+        The predictions for this fold.
+    """
+
+    scores: list[ScoreResult]
+    ground_truth: Optional[pd.Series] = None
+    predictions: Optional[pd.Series] = None
+    train_data: Optional[pd.Series] = None
+
+
+@dataclass
+class ResultObject:
+    """
+    Model results for a single task.
+
+    Parameters
+    ----------
+    model_id : str
+        The ID of the model.
+    task_id : str
+        The ID of the task.
+    folds : list of FoldResults
+        The results for each fold.
+    means : list of ScoreResult
+        The mean scores across all folds for each scorer.
+    stds : list of ScoreResult
+        The standard deviation of scores across all folds for
+        each scorer.
+    """
+
+    model_id: str
+    task_id: str
+    folds: dict[int, FoldResults]
+    means: list[ScoreResult] = field(init=False)
+    stds: list[ScoreResult] = field(init=False)
+
+    def __post_init__(self):
+        """Calculate mean and std for each score."""
+        self.means = []
+        self.stds = []
+        scores = {}
+        for fold_idx, fold in self.folds.items():
+            for score in fold.scores:
+                if score.name not in scores:
+                    scores[score.name] = []
+                scores[score.name].append(score.score)
+        for name, score in scores.items():
+            # TODO mean wrongly calculated over all axis.
+            self.means.append(ScoreResult(name, np.mean(score, axis=0)))
+            self.stds.append(ScoreResult(name, np.std(score, axis=0)))
+
+
+@dataclass
+class BenchmarkingResults:
+    """Results of a benchmarking run.
+
+    Parameters
+    ----------
+    results : list of ResultObject
```

Comment:
results are indexed by task and model ID, so a `dict` would be far more efficient for retrieval. `contains` would then just be a query of the keys.

Reply:
As also mentioned somewhere below. Having a class enables us to define methods on the object and thus the code is less complicated. (E.g. to_dataframe here).